astrolabe/validator.py

```diff
@@ -106,194 +93,395 @@ def run_test_expecting_error(self, driver_workload):
+
+        if stats['numErrors'] < 0:
+            self.fail_stats(
+                "Expected numErrors to be non-negative, but got {} instead."
+                .format(stats['numErrors']))
+
+        if stats['numFailures'] < 0:
+            self.fail_stats(
+                "Expected numFailures to be non-negative, but got {} instead."
+                .format(stats['numFailures']))
+
+    # Assert contents of the events.json file. The presence and type of each
+    # required field (i.e. events, errors, failures) will always be asserted.
+    # Based on the corresponding parameter for each field, the function will
+    # then assert either that the array is empty or that it is non-empty and
+    # each document therein has an expected structure.
+    #
+    # By default, this function will assert that events.json contains some
+    # events but no errors or failures.
+    #
+    # If hasErrorsXorFailures is true, hasErrors and hasFailures will be ignored
+    # and the function will instead assert that either errors or failures are
+    # present (but not both).
+    def assert_events(self, hasEvents=True, hasErrors=False, hasFailures=False,
+                      hasErrorsXorFailures=False):
+        try:
+            events = yaml.safe_load(open('events.json').read())
+        except OSError as exc:
+            self.fail("Failed to open events.json: %s" % exc)
+        except yaml.YAMLError as e:
+            self.fail("Failed to parse events.json: %s" % e)
+
+        # Assert both presence and type of required fields in events.json
+        self.assert_has_keys(events, ['events', 'errors', 'failures'])
+
+        if not isinstance(events['events'], list):
+            self.fail("The workload executor didn't record events as an array.")
+
+        if not isinstance(events['errors'], list):
+            self.fail("The workload executor didn't record errors as an array.")
+
+        if not isinstance(events['failures'], list):
+            self.fail("The workload executor didn't record failrues as an array.")
+
+        # If hasEvents is true, assert that the array is non-empty and that each
+        # object within contains essential fields. If hasEvents is false, the
+        # array should be empty.
+        #
+        # Note: we do not assert the type of events observed since not all
+        # drivers implement CMAP and thus might only log Command events.
+        if hasEvents:
+            if not events['events']:
+                self.fail("The workload executor recorded no events but some were expected")
+
+            numConnectionEvents = 0
+            numPoolEvents = 0
+            numCommandEvents = 0
+
+            for event in events['events']:
+                if ('name' not in event) or (not event['name'].endswith('Event')):
+                    self.fail("The workload executor didn't record event name as expected.")
+
+                if 'observedAt' not in event:
```

Comment:
I was originally asserting that this was a float but later realized that was unreliable. For example, Java produces `1.618408012349E9` as a numeric literal in its `events.json`, which `yaml.safe_load` parses as a string.